buildSrc/subprojects/parser/src/main/kotlin/org/apache/calcite/buildtools/parser/DialectGenerateTask.kt

```diff
@@ -0,0 +1,210 @@
+    private fun extractFunctions(): Map<String, String> {
+        val rootDirectoryFile = rootDirectory.get().asFile
+        val queue = getTraversalPath(rootDirectoryFile)
+        val functionMap: MutableMap<String, String> = HashMap<String, String>()
+        traverse(queue, rootDirectoryFile, functionMap)
+        return functionMap
+    }
+
+    /**
+     * Generates the parserImpls.ftl file for the dialect.
+     *
+     * @param functions The functions to place into the output file
+     */
+    private fun generateParserImpls(functions: Map<String, String>) {}
+
+    /**
+     * Gets the traversal path for the dialect by "subtracting" the root
+     * absolute path from the dialect directory absolute path.
+     *
+     * @param rootDirectoryFile The file for the root parsing directory
+     */
+    private fun getTraversalPath(rootDirectoryFile: File): Queue<String> {
+        val dialectDirectoryFile = dialectDirectory.get().asFile
+        var dialectPath = dialectDirectoryFile.absolutePath
+        val rootPath = rootDirectoryFile.absolutePath
+        val rootIndex = dialectPath.indexOf(rootPath)
+        dialectPath = dialectPath.substring(rootIndex + rootPath.length + 1)
+        return LinkedList(dialectPath.split("/"))
+    }
+
+    /**
+     * Traverses the determined path given by the queue. Once the queue is
+     * empty, the dialect directory has been reached. In that case any *.ftl
+     * file should be processed and no further traversal should happen.
+     *
+     * @param directories The directories to traverse in topdown order
+     * @param currentDirectory The current directory the function is processing
+     * @param functionMap The map to which the parsing functions will be added to
+     */
+    private fun traverse(
+        directories: Queue<String>,
+        currentDirectory: File,
+        functionMap: MutableMap<String, String>
+    ) {
+        val files = currentDirectory.listFiles()
+        // Ensures that files are processed first.
+        files.sortBy { it.isDirectory }
+        val nextDirectory = directories.peek()
+        for (f in files) {
+            if (f.isFile && f.extension == "ftl") {
+                processFile(f, functionMap)
+            }
+            if (directories.isNotEmpty() && f.name == nextDirectory) {
+                directories.poll()
+                traverse(directories, f, functionMap)
+            }
+        }
+    }
+
+    /**
+     * Extracts the functions from the given file into functionMap. Parses
+     * functions of the form:
+     * <return_type> <name>(<args>) :
+     * {
+     *     <properties>
+     * }
+     * {
+     *     <body>
+     * }
+     * @param file The file to process
+     * @param functionMap The map to which the parsing functions will be added to
+     */
+    private fun processFile(file: File, functionMap: MutableMap<String, String>) {
+        var fileText = file.readText(Charsets.UTF_8)
+        val typeAndName = "\\w+\\s+\\w+"
```

Comment:
Creating regex objects can be pretty expensive. Rather than defining/initializing typeAndName, declarationPattern, delims, and splitRegex locally for every file, can we just create the equivalent of a global version of each? Alternatively, can we create them in a higher level function and pass them in here?

Reply:
Yep, I can move them to be global.